### DISCUSSION TITLE
Use type inference for the Depends params function return type

### DISCUSSION THREAD

**Discussion by jvllmr** [fastapi/fastapi#9338] [Questions]
### First Check

- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I added a very descriptive title here.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import AsyncGenerator, Generator, Literal
from fastapi import Depends

# type of a is Any
a = Depends()


def b_() -> Literal[1]:
    return 1

# type of b is Literal[1] with type inference, Any without
b = Depends(b_)


def c_() -> Generator[Literal[1], None, None]:
    yield 1

# type of c is Literal[1] with type inference, Any without
c = Depends(c_)


async def d_() -> Literal[1]:
    return 1

# type of d is Literal[1] with type inference, Any without
d = Depends(d_)


async def e_() -> AsyncGenerator[Literal[1], None]:
    yield 1

# type of e is Literal[1] with type inference, Any without
e = Depends(e_)
```

### Description

First of all, a big thanks to the FastAPI team for this amazing project. I just recently started using FastAPI and backend-development is fun like never before for me. However, there is still one pain point for me when using FastAPI: Having to declare a type every time I use Depends. For that reason I created a wrapper in my project which uses the following type overloads (imports not included):

```python
_TDepends = TypeVar("_TDepends")


@overload
def Depends(
    dependency: Callable[..., AsyncGenerator[_TDepends, None]],
    *,
    use_cache: bool = True,
) -> _TDepends:
    ...


@overload
def Depends(
    dependency: Callable[..., Coroutine[None, None, _TDepends]],
    *,
    use_cache: bool = True,
) -> _TDepends:
    ...


@overload
def Depends(
    dependency: Callable[..., Generator[_TDepends, None, None]],
    *,
    use_cache: bool = True,
) -> _TDepends:
    ...


@overload
def Depends(
    dependency: Callable[..., _TDepends], *, use_cache: bool = True
) -> _TDepends:
    ...


@overload
def Depends(dependency: None = None, *, use_cache: bool = True) -> Any:
    ...


def Depends(
    dependency: Optional[Callable[..., Any]] = None,
    *,
    use_cache: bool = True,
) -> Any:  # noqa: N802
    return params.Depends(dependency=dependency, use_cache=use_cache)
```

This allows me to use Depends without having to declare the return type every time, which saves me a lot of boilerplate. I think it would be great if FastAPI would support this behaviour out of the box. 

There is one only issue I see with this: The inferred return type only makes sense within FastAPI's dependency system and if anyone wants to use this function anywhere else the inferred return type creates problems. After putting a lot of thought into this I realized that people should use the actual Depends class in these cases anyway.

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

3.11

### Additional Context

I already created a fork and put my type overloads there: https://github.com/jvllmr/fastapi

**Comment by YuriiMotov** [comment]
It's recommended now to use the approach with Annotated:
```py
@router.post("/login")
def login(
  database: Annotated[DBSession, Depends(get_database)],
):
  pass
```

Generic `Depends` will not help when `Annotated` approach is used.

There are a couple of PR that aimed to implement this, but thy were closed with that reason:
* https://github.com/fastapi/fastapi/pull/11255
* https://github.com/fastapi/fastapi/pull/10053
